Feat: LIP-33 - GateSeal v2.0 (#85)
lidofinance/lido-improvement-proposals: LIPS/lip-32.md

---
lip: 32
title: GateSeal V2.0 — Streamlining one-time emergency buttons
status: WIP
author: Alexey Komarov
discussions-to: <https://research.lido.com/t/gateseal-v2-0-proposal-thread>

created: 2025-06-04
requires: 
implementation: 
---

## Simple Summary

GateSeal V2.0 introduces a more flexible, long-lived version of Lido’s emergency circuit breaker. It aims to preserve one-time emergency protection while drastically reducing governance and operational overhead by enabling on-chain expiry prolongation and modernizing the codebase.

## Abstract

> We propose to deploy a new version of the GateSeal contract that adds:

Reviewer: This implies a single GateSeal contract. Consider rephrasing like "upgrading GateSeal blueprint and factory that includes".

- configurable initial expiry;
- on-chain expiry prolongations;
- committee-owned prolongations (limited count);
- pre-deployed, module-specific contracts;
- strict signing schedule.
The new implementation will be written in Vyper 0.4.2, improving security, auditability and expressiveness while maintaining backward compatibility with existing PausableUntil-based contracts.

## Motivation

> GateSeal V1 has never been triggered but has been repeatedly re-deployed due to its hardcoded 1-year expiry. This creates governance overhead and operational burden for the Lido DAO. Its inability to prolong lifespan without full redeployment exposes the DAO to unnecessary coordination cycles and short renewal windows. A more sustainable, long-lived model is needed to preserve the utility of the emergency switch without recurring maintenance.

Reviewer: This V1 governance overhead was intended by design, as gateseals were conceived as a simple temporary solution, and this overhead was supposed to motivate DAO to come up with a longterm sustainable solution. Perhaps, it's worth mentioning because V2 is a shift in design philosophy.

> GateSeal V1 has never been triggered but has been repeatedly re-deployed due to its hardcoded 1-year expiry. This creates governance overhead and operational burden for the Lido DAO. Its inability to prolong lifespan without full redeployment exposes the DAO to unnecessary coordination cycles and short renewal windows. A more sustainable, long-lived model is needed to preserve the utility of the emergency switch without recurring maintenance.

Reviewer: Perhaps needs a link to GateSeal V1 for context

Author: added the [link](https://github.com/lidofinance/gate-seals/blob/564955d1ec30ee832179ebcef131392a5fe8ab69/README.md?plain=1#L22)

## Specification

### Glossary

- **Prolongation Window** - the active window during which the committee can prolong the contract.
- **Prolongation Period** — the extra time added to the contract with each prolongation.
- **Initial Lifetime** — the time from deployment until the first expiration time.
- **Prolongation Count** — the maximum number of allowed prolongations.

> - **DAO Ops Reserve** — a 60-day buffer for DAO Ops to deploy a new GateSeal before the current one expires.

Reviewer: Needs a short rationale for why exactly this period (60 days) was decided upon

### Rationale

> The main tradeoff considered was between maintaining operational safety and reducing coordination overhead. Adding a prolongation mechanism shifts execution of the validity renewal from the DAO to the emergency committee, enforcing drills through real multisig action. The calculated maximum number of prolongations and enforcing liveness checks maintains a balance between flexibility and rigorness. Vyper 0.4.2 brings substantial improvements in security, gas efficiency, and maintainability by fixing known compiler issues (e.g. [GHSA-w9g2-3w7p-72g9 advisory](https://github.com/advisories/GHSA-w9g2-3w7p-72g9)), adding full support for the EVM Cancun hard fork, and refining gas-cost optimizations. For the full changelog, refer to the [official release notes](https://docs.vyperlang.org/en/stable/release-notes.html#v0-4-2-lernaean-hydra). The model stays true to the one-time-use principle to ensure GateSeal remains a high-trust, high-scrutiny tool.

Reviewer: > rigorness

rigorousness, strictness

Author: `rigor`

  - number of **sealables** is zero, too many, or contains duplicates
  - calculated total lifetime exceeds the allowed 5-year cap

- Triggering `seal()`:
  - pauses all configured sealables
  - emits a `Sealed` event with correct parameters
  - causes the GateSeal to become expired

- `prolongLifetime()`:
  - succeeds only when called within the configured prolongation window
  - extends the GateSeal's lifetime by **Prolongation Period**
  - decrements the number of prolongations remaining
  - emits a `Prolonged` event with updated values

- `prolongLifetime()` reverts if:
  - called before the prolongation window opens
  - called after the window closes
  - the GateSeal is already sealed (and thus expired)
  - the GateSeal has no prolongations left
  - called by a non-committee account

- `seal()` and `prolongLifetime()` can only be called by the committee; all unauthorized calls revert
- A second call to `prolongLifetime()` within the same window reverts, enforcing one action per window

## Security Considerations

- GateSeal V2 must remain single-use.
- Prolongation is only possible for unused and unexpired contracts, and must be performed within the prolongation window.

> - Signature-based prolongation actions double as committee drills. Only unused and unexpired contracts may be prolonged, and only within the prolongation window.

Reviewer: Every action is signature-based. Don't think we need to specify this. Might lead to misunderstanding

---
lip: 32
title: GateSeal V2 — Streamlining one-time emergency buttons
status: WIP
author: Alexey Komarov
discussions-to: <https://research.lido.com/t/gateseal-v2-0-proposal-thread>

created: 2025-06-04
requires: 
implementation: 
---

## Simple Summary

> GateSeal V2 introduces a more flexible, long-lived version of Lido’s emergency circuit breaker. It aims to preserve one-time emergency protection while drastically reducing governance and operational overhead by enabling on-chain expiry prolongation and modernizing the codebase.

Reviewer 2: ```suggestion
GateSeal V2 introduces a more flexible, long-lasting version of Lido’s emergency circuit breaker. It aims to preserve single-use emergency protection while drastically reducing governance and operational overhead by enabling on-chain expiry prolongation and modernizing the codebase.
```

## Abstract

We propose to upgrade the GateSeal blueprint and factory to include:
- configurable initial expiry;
- on-chain expiry prolongations;
- committee-owned prolongations (limited count);

> - pre-deployed, module-specific contracts;

Reviewer 2: the contracts are not module-specific, the sealables are **organized around** or **grouped by** each module

- strict signing schedule.

The new implementation will be written in Vyper 0.4.2, improving security, auditability and expressiveness while maintaining backward compatibility with existing PausableUntil-based contracts.

## Motivation

[GateSeal V1 served as a temporary mechanism](https://github.com/lidofinance/gate-seals/blob/564955d1ec30ee832179ebcef131392a5fe8ab69/README.md?plain=1#L22) that intentionally introduced operational overhead to prompt the DAO to develop a long-term sustainable solution. Now that GateSeal has become a permanent part of Lido’s emergency infrastructure, its architecture must evolve — from a short-term, incentive-based tool to a reliable and maintainable mechanism.

## Specification

### Glossary

> - **Prolongation Window** - the active window during which the committee can prolong the contract.

Reviewer 2: ```suggestion
- **Prolongation Window** — the period during which the committee can prolong the contract.
```

- **Prolongation Period** — the extra time added to the contract with each prolongation.
- **Initial Lifetime** — the time from deployment until the first expiration time.
- **Prolongation Count** — the maximum number of allowed prolongations.
- **DAO Ops Reserve** — a 60-day buffer that ensures the protocol remains protected if the committee fails to prolong in time, giving DAO Ops enough time to coordinate a planned replacement. The buffer covers both the time required to form a new multisig committee (2–4 weeks) and to conduct a full-fledged Dual Governance vote (2–4 weeks).

### Rationale

> The main tradeoff considered was between maintaining operational safety and reducing coordination overhead. Adding a prolongation mechanism shifts the execution of validity renewals from the DAO to the emergency committee, enforcing drills through real multisig action. The calculated maximum number of prolongations and the enforcement of liveness checks maintain a balance between flexibility and rigor. Vyper 0.4.2 brings substantial improvements in security, gas efficiency, and maintainability by fixing known compiler issues (e.g. [GHSA-w9g2-3w7p-72g9 advisory](https://github.com/advisories/GHSA-w9g2-3w7p-72g9)), adding full support for the EVM Cancun hard fork, and refining gas-cost optimizations. For the full changelog, refer to the [official release notes](https://docs.vyperlang.org/en/stable/release-notes.html#v0-4-2-lernaean-hydra). The model stays true to the one-time-use principle to ensure GateSeal remains a high-trust, high-scrutiny tool.

Reviewer 2: ```suggestion
The main tradeoff considered was between maintaining operational safety and reducing coordination overhead. Adding a prolongation mechanism shifts the execution of validity renewals from the DAO to the emergency committee, enforcing drills through real multisig actions. The calculated maximum number of prolongations and the enforcement of liveness checks maintain a balance between flexibility and rigor. Vyper 0.4.2 brings substantial improvements in security, gas efficiency, and maintainability by fixing known compiler issues (e.g. [GHSA-w9g2-3w7p-72g9 advisory](https://github.com/advisories/GHSA-w9g2-3w7p-72g9)), adding full support for the EVM Cancun hard fork, and refining gas-cost optimizations. For the full changelog, refer to the [official release notes](https://docs.vyperlang.org/en/stable/release-notes.html#v0-4-2-lernaean-hydra). The model maintains the single-use principle to ensure GateSeal remains a high-trust, high-scrutiny tool.
```

### Technical Specification

- The contract will be written in Vyper 0.4.2.
- `prolongLifetime()` callable only by the committee when unused, unexpired, within the prolongation window, and with remaining prolongations.
- Each call prolongs the lifetime by the **Prolongation Period** and decrements remaining prolongations.
- `seal()` pauses configured contracts for **SEAL_DURATION_SECONDS** and expires the GateSeal immediately.
- **Prolongation Period** is set to **1 year**, ensuring that the committee needs to act only once per contract per year.
- [**Prolongation Window**](https://www.notion.so/ADR-X-GateSeal-v2-216bf633d0c9809994ebd484c6334e42?source=copy_link#216bf633d0c9802a870ce30b883f2cad) is fixed at **14 days** — as this remains an emergency mechanism, the committee is expected to be available for two planned sessions per year.
- **Initial Lifetime** is set at deployment and must not exceed `2 × Prolongation Period` (**2 years**), while also respecting a lower bound defined by `DAO Ops Reserve + Prolongation Window` (**74 days**), to guarantee a safe buffer before the first prolongation.

> - **Prolongation Count** is defined at deployment, enabling custom total lifetimes per contract within global constraints.

Reviewer 2: Which global constraints?

- **Maximum Total Lifetime** — calculated as `Initial Lifetime + (Prolongation Period × Prolongation Count)` — must not exceed **5 years**; this is enforced at deployment time.

> - The ability to **set custom seal duration limits at deployment** is removed — prior limits (e.g., 4–14 days) are now obsolete. Deployment-time configuration is immutable and subject to deploy-time verification, which provides sufficient safeguards.

Reviewer 2: Why the prior limits are now obsolete?

---
lip: 32

> title: GateSeal V2.0 – Streamlining Emergency Governance

Reviewer 3: ```suggestion
title: GateSeal V2.0 — Streamlining one-time emergency buttons
```

status: WIP
author: Alexey Komarov
discussions-to: <https://research.lido.com/t/gateseal-v2-0-proposal-thread>

created: 2025-06-04
requires: 
implementation: 
---

## Simple Summary

GateSeal V2.0 introduces a more flexible, long-lived version of Lido’s emergency circuit breaker. It aims to preserve one-time emergency protection while drastically reducing governance and operational overhead by enabling on-chain expiry prolongation and modernizing the codebase.

## Abstract

> We propose to deploy a new version of the GateSeal contract that will support a configurable initial expiry of up to 3 years and allow a limited number of on-chain expiry prolongations. Each prolongation may increase the contract's lifetime by up to 6 months, subject to conditions proving the committee’s liveness. The maximum pause window will also be increased from 14 to 21 days. The new implementation will be written in Vyper 0.4.1, improving security and gas efficiency while maintaining backward compatibility with existing PausableUntil-based contracts.

Reviewer 3: ```suggestion
We propose to deploy a new version of the GateSeal contract that will support a configurable initial expiry of up to 3 years and allow a limited number of on-chain expiry prolongations. Each prolongation may increase the contract's lifetime by up to 6 months, subject to conditions proving the committee’s liveness. The maximum pause window will also be increased from 14 to 21 days. The new implementation will be written in Vyper 0.4.1, improving security, auditability and expressiveness while maintaining backward compatibility with existing PausableUntil-based contracts.
```

## Motivation

> GateSeal V1 has never been triggered but has been repeatedly re-deployed due to its hardcoded 1-year expiry. This creates governance overhead and operational burden for the emergency committee. Its inability to prolong lifespan without full redeployment exposes the DAO to unnecessary coordination cycles and short renewal windows. A more sustainable, long-lived model is needed to preserve the utility of the emergency switch without recurring maintenance.

Reviewer 3: ```suggestion
GateSeal V1 has never been triggered but has been repeatedly re-deployed due to its hardcoded 1-year expiry. This creates governance overhead and operational burden for the Lido DAO. Its inability to prolong lifespan without full redeployment exposes the DAO to unnecessary coordination cycles and short renewal windows. A more sustainable, long-lived model is needed to preserve the utility of the emergency switch without recurring maintenance.
```

## Specification

### Overview

GateSeal V2 will retain the one-time activation logic but allow:

> - An initial expiry up to 3 years (set at deployment),

Reviewer 3: ```suggestion
- An initial expiry up to 1 year (set at deployment),
- An overall expiry up to 5 years (set at deployment),
```

Reviewer 4: We have N prolongations, so why do we need a cap in the contract? It seems like an extra precaution.

Reviewer 3: want to limit the overall duration of the same instance still to avoid some possibly forgotten stuff

> - Up to N prolongations (e.g., 5), each increasing validity by up to 6 months,

Reviewer 4: ```suggestion
- Up to N prolongations (e.g., 5), each increasing validity by up to 6 months (set at deployment),
```

- Prolongations callable only if the seal is unused, unexpired, and has remaining credit,
- A new pause duration limit of up to 21 days,
- Deployment using Vyper 0.4.1.

### Rationale

> The main tradeoff considered was between maintaining operational safety and reducing coordination overhead. Adding a prolongation mechanism shifts responsibility for validity renewal from the DAO to the committee, enforcing drills through real multisig action. Limiting prolongations and enforcing liveness checks maintains a balance between flexibility and rigor. The switch to Vyper 0.4.1 removes compiler-level bugs and increases auditability. The model stays true to the one-time-use principle to ensure GateSeal remains a high-trust, high-scrutiny tool.

Reviewer 3: ```suggestion
The main tradeoff considered was between maintaining operational safety and reducing coordination overhead. Adding a prolongation mechanism shifts execution of the validity renewal from the DAO to the emergency committee, enforcing drills through real multisig action. Limiting prolongations and enforcing liveness checks maintains a balance between flexibility and rigorness. Vyper 0.4.1 brings substantial improvements in security, gas efficiency, and maintainability by fixing known compiler issues, adding full support for the EVM Cancun hard fork, and refining gas-cost optimizations. For the full changelog, refer to the [official release notes](https://docs.vyperlang.org/en/stable/release-notes.html#v0-4-1-tokara-habu). The model stays true to the one-time-use principle to ensure GateSeal remains a high-trust, high-scrutiny tool.
```

### Technical Specification

- Contract will be written in Vyper ≥0.4.1.

> - Public method `prolong()` will be gated by internal checks:

Reviewer 3: ```suggestion
- Public method `extendLifetime()` will be gated by internal checks:
```

  - `not activated`
  - `not expired`

>   - `prolongations_remaining > 0`

Reviewer 3: ```suggestion
  - `prolongations_remaining > 0`
  - can't be called more than once for the current period
  - can't be called earlier than 1 mo before the currently assigned expiration
```

- The `pause()` method behaves identically to V1 and triggers self-destruction after invocation.
- Max `pause_until` will increase from 14 to 21 days.
- Full compatibility with existing PausableUntil contracts (e.g., WithdrawalQueue, ValidatorExitBusOracle).

### Test Cases

- `pause()` is called → GateSeal self-destructs → state is frozen for up to 21 days.
- `prolong()` is called before expiry → expiry timestamp increases.
- `prolong()` fails if:
  - Contract is expired,
  - Already paused,
  - No remaining prolongations.

## Security Considerations

- GateSeal V2 must remain single-use.
- Only unused and unexpired contracts may be prolonged.
- Limited number of prolongations ensures eventual expiry.
- Signature-based prolongation actions double as committee drills.
- Increased maximum pause duration still fits within dual governance emergency limits.

## Failure Modes

- If the committee becomes inactive and all prolongations are used, the GateSeal will expire, requiring a redeploy.
- Incorrectly configured parameters (e.g., excessive max prolongations or duration) may reduce accountability.

> - Misuse of the prolongation function is mitigated by strict internal checks and public transparency of calls.

Reviewer 3: ```suggestion
- Misuse of the prolongation function is mitigated by strict internal checks and public transparency of calls from the reference deployment factory.
```
